rehttp_server_post17_test.go

```diff
@@ -0,0 +1,167 @@
+//go:build go1.7
+// +build go1.7
+
+package rehttp
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+)
+
+func TestTransport_RoundTripTimeouts(t *testing.T) {
+	// to keep track of any open server requests and ensure the correct number of requests were made
+	ch := make(chan bool, 4)
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ch <- true
+		time.Sleep(time.Millisecond * 100)
+		w.WriteHeader(http.StatusTooManyRequests)
+	}))
+	defer ts.Close()
+
+	tr := NewTransport(
+		http.DefaultTransport,
+		RetryAll(RetryMaxRetries(3), RetryAny(
+			RetryStatuses(http.StatusTooManyRequests), // retry 429s
+			func(attempt Attempt) bool { // retry context deadline exceeded errors
+				return attempt.Error != nil && attempt.Error == context.DeadlineExceeded // errors.Is requires go 1.13+
+			})),
+		ConstDelay(0),
+	)
+	tr.PerAttemptTimeout = time.Millisecond * 10 // short timeout
+
+	client := http.Client{
+		Transport: tr,
+	}
+
+	req, err := http.NewRequest(http.MethodGet, ts.URL, nil)
+	if err != nil {
+		t.Errorf("error creating request: %s", err)
+	}
+
+	_, err = client.Do(req)
+	// make sure server has finished and got 4 attempts
+	<-ch
+	<-ch
+	<-ch
+	<-ch
+	if err == nil {
+		t.Error("expected timeout error doing request but got nil")
+	}
+
+	// now increase the timeout restriction
+	ch = make(chan bool, 4)
+	tr.PerAttemptTimeout = time.Second
+	res, err := client.Do(req)
+
+	// should have attempted 4 times without going over the timeout
+	<-ch
+	<-ch
+	<-ch
+	<-ch
+
+	if err != nil {
+		t.Errorf("got unexpected error doing request: %s", err)
+	}
+	if res == nil || res.StatusCode != http.StatusTooManyRequests {
+		t.Errorf("status code does not match expected: got %d, want %d", res.StatusCode, http.StatusTooManyRequests)
```

Comment:
This will panic if `res` is `nil`. I suggest calling `t.Fatalf` on line 68 above, so that after this condition `res` is necessarily not nil. (I can take care of this change when I implement the channel thing I mentioned above).